iOS_SDK/OneSignalSDK/Source/OneSignal.m

```diff
@@ -2613,28 +2619,40 @@ + (void)setExternalUserId:(NSString * _Nonnull)externalId {
 }
 
 + (void)setExternalUserId:(NSString * _Nonnull)externalId withCompletion:(OSUpdateExternalUserIdBlock _Nullable)completionBlock {
-    
     // return if the user has not granted privacy permissions
     if ([self shouldLogMissingPrivacyConsentErrorWithMethodName:@"setExternalUserId:withCompletion:"])
         return;
+    
+    [self setExternalUserId:externalId withExternalIdAuthHashToken:nil withCompletion:completionBlock];
+}
+
++ (void)setExternalUserId:(NSString *)externalId withExternalIdAuthHashToken:(NSString *)hashToken withCompletion:(OSUpdateExternalUserIdBlock)completionBlock {
+    
+    // return if the user has not granted privacy permissions
+    if ([self shouldLogMissingPrivacyConsentErrorWithMethodName:@"setExternalUserId:withExternalIdAuthHashToken:withCompletion:"])
+        return;
 
     // Can't set the external id if init is not done or the app id or user id has not ben set yet
     if (!performedOnSessionRequest) {
         // will be sent as part of the registration/on_session request
         pendingExternalUserId = externalId;
+        pendingExternalUserIdHashToken = hashToken;
         return;
     } else if (!self.currentSubscriptionState.userId || !self.app_id) {
         [OneSignal onesignal_Log:ONE_S_LL_WARN message:[NSString stringWithFormat:@"Attempted to set external user id, but %@ is not set", self.app_id == nil ? @"app_id" : @"user_id"]];
         return;
+    } else if (self.currentEmailSubscriptionState.requiresEmailAuth && !hashToken) {
```

Comment:
`self.currentEmailSubscriptionState.requiresEmailAuth` should be checking `requiresUserIdAuth` instead